djconnectwise/sync.py

```diff
@@ -1596,6 +1599,75 @@ def __init__(self, *args, **kwargs):
             self.batch_condition_list = \
                 list(filtered_statuses.values_list('id', flat=True))
 
+    def get_batch_condition(self, conditions):
+        return 'status/id in ({})'.format(
+            ','.join([str(i) for i in conditions])
+        )
+
+    def get_page(self, *args, **kwargs):
+        return self.client.get_tickets(*args, **kwargs)
+
+    def get_single(self, ticket_id):
+        return self.client.get_ticket(ticket_id)
+
+    def sync_related(self, instance):
+        instance_id = instance.id
+        sync_classes = []
+
+        note_sync = ServiceNoteSynchronizer()
+        note_sync.api_conditions = [instance_id]
+        sync_classes.append((note_sync, Q(ticket=instance)))
+
+        time_sync = TimeEntrySynchronizer()
+        time_sync.batch_condition_list = [instance_id]
+        sync_classes.append((time_sync, Q(charge_to_id=instance)))
+
+        activity_sync = ActivitySynchronizer()
+        activity_sync.api_conditions = ['ticket/id={}'.format(instance_id)]
+        sync_classes.append((activity_sync, Q(ticket=instance)))
+
+        self.sync_children(*sync_classes)
+
+    def fetch_sync_by_id(self, instance_id):
+        instance = super().fetch_sync_by_id(instance_id)
+        if not instance.closed_flag:
+            self.sync_related(instance)
+        return instance
+
+    def _instance_ids(self, filter_params=None):
+        tickets_qset = self.filter_by_record_type()
+
+        if not filter_params:
+            ids = tickets_qset.values_list(self.lookup_key, flat=True)
+        else:
+            ids = tickets_qset.filter(filter_params).values_list(
+                self.lookup_key, flat=True
+            )
+        return set(ids)
+
+    def get_delete_qset(self, stale_ids):
+        tickets = self.filter_by_record_type()
+        return tickets.filter(pk__in=stale_ids)
+
+    def get_sync_job_qset(self):
+        return models.SyncJob.objects.filter(
+            entity_name=self.model_class.__name__,
+            synchronizer_class=self.__class__.__name__
+        )
+
+
+class TicketSynchronizer(TicketSynchronizerMixin,
```

Comment:
You should change TicketSynchronizer to ServiceTicketSynchronizer. It seems to be inconsistent in some places, and technically some places it has to be ambiguous because it refers to both models. For that reason especially, I think we should tighten up our terminology.

I'm sorry to say this, because it will probably be annoying to do, but I think it will be better in the long run.

Reply:
Yes. This was a good suggestion. It uncovered the fact that I wasn't actually handling making sync requests to the API using the new project synchronizer. So I've still got some changes to make for callbacks and in card_detail. At the moment the `ConnectWiseServiceDetailView` will make a request with the ticket synchronizer if the ticket doesn't exist in the DB already. This should be pretty rare but it means another place where we've got to choose which synchronizer to use.